engine/consensus/matching/engine.go

```diff
@@ -131,24 +128,69 @@ func (e *Engine) Process(originID flow.Identifier, event interface{}) error {
 func (e *Engine) HandleReceipt(originID flow.Identifier, receipt flow.Entity) {
 	e.log.Debug().Msg("received receipt from requester engine")
 	e.metrics.MessageReceived(metrics.EngineSealing, metrics.MessageExecutionReceipt)
-	e.pendingReceipts.Put(&engine.Message{OriginID: originID, Payload: receipt})
+	e.pendingReceipts.Push(receipt)
 	e.notifier.Notify()
 }
 
 // OnFinalizedBlock implements the `OnFinalizedBlock` callback from the `hotstuff.FinalizationConsumer`
 // CAUTION: the input to this callback is treated as trusted; precautions should be taken that messages
 // from external nodes cannot be considered as inputs to this function
 func (e *Engine) OnFinalizedBlock(finalizedBlockID flow.Identifier) {
-	e.pendingFinalizationEvents.Push(finalizedBlockID)
-	e.notifier.Notify()
+	e.finalizationEventsNotifier.Notify()
+	e.processFinalizedReceipts(finalizedBlockID)
+}
+
+// processFinalizedReceipts selects receipts that were included into finalized block and submits them
+// for further processing by matching core.
+// Without the logic below, the sealing engine would produce IncorporatedResults
+// only from receipts received directly from ENs. sealing Core would not know about
+// Receipts that are incorporated by other nodes in their blocks blocks (but never
+// received directly from the EN).
+func (e *Engine) processFinalizedReceipts(finalizedBlockID flow.Identifier) {
```

Comment:
Some comments regarding this method:

### (1) The problem this logic is supposed to address

Sorry, I might have given misleading directions on this particular detail. Your implementation helped me consolidate my thoughts:
* In the previous implementation, we only consumed receipts directly from Execution Nodes. 
* We have logic in `matching.Core` (-> [code](https://github.com/onflow/flow-go/blob/b7df2882af2a49525c8f3a901d1c02a6a474fcd1/engine/consensus/matching/core.go#L266)) for re-requesting receipts, which got lost. Therefore, the logic for consuming receipts from blocks is also **_not_ critical for liveness**

So why did I ask for this logic to be added:
* consistency: when the node has learned about the existence of a receipt being included in a block, the sealing logic should also consider this receipt
* performance: re-requesting a receipt we already know about adds latency and consumes networking resources 

#### Consequences:
1. This logic can work on a best-effort basis; we should _not be crashing the node_ when we fail to queue a receipt: https://github.com/onflow/flow-go/blob/b7df2882af2a49525c8f3a901d1c02a6a474fcd1/engine/consensus/matching/engine.go#L170-L175 (as the receipt will eventually be re-requested) 
2. We want to add the receipt _as soon as we learn about its existence_ to increase responsiveness. Hence, this logic should inspect blocks when they are first incorporated (HotStuff notification: `OnBlockIncorporated`). Don't wait until the block is finalized. 

### (2) Including the logic into the `matching.Engine`

We generally queue inbound work for an engine and have a limited number of routines crunch   through that work.  I would prefer if we applied the same logic here:
* When catching up, a node might receive incorporate hundreds of blocks a second. Spawning many go routines here might deprive other components of resources, when the node is behind. Hence, I think queuing the incorporated blocks would be favourable and assign a single worker to the queue. 
* The queue for freshly incorporated blocks can be quite short (maybe 10). If the node is keeping up there should only ever be one or two blocks in the queue. If the node is behind, the receipts in the blocks are also likely already sealed. Hence, dropping blocks is a good option. 


### (3) Specific implementation: 

* We retrieve the entire block from storage, which adds some GC load as we are re-assembling the entire payload from its components on each call. But we are only interested in the receipts. You can get the receiptIDs from `storage.Index`.  
* You re-assemble the receipt from its meta-information and the result. You also have a lookup (i.e. a mini cache) for the results that are in the block. All this logic is already implemented by `storage.ExecutionReceipts` 

#### Suggestion

* instead of https://github.com/onflow/flow-go/blob/b7df2882af2a49525c8f3a901d1c02a6a474fcd1/engine/consensus/matching/engine.go#L29-L30 you could use 
  ```golang
  receipts                   storage.ExecutionReceipts
  index                      storage.Index
  ```
* This is all the logic you need here:
  ```golang
  func (e *Engine) processFinalizedReceipts(finalizedBlockID flow.Identifier) {
  	e.unit.Launch(func() {
  		index, err := e.index.ByBlockID(finalizedBlockID)
  		if err != nil {
  			e.log.Fatal().Err(err).Msgf("could not retrieve payload index for block %v", finalizedBlockID)
  		}

  		for _, receiptID := range index.ReceiptIDs {
  			receipt, err := e.receipts.ByID(receiptID)
  			if err != nil {
  				e.log.Fatal().Err(err).Msgf("could not retrieve receipt incorporated in block %v", finalizedBlockID)
  			}
  			e.pendingReceipts.Push(receipt)
  		}
  		e.notifier.Notify()
  	})
  }
  ```

Reply:
I have implemented requested changes for matching engine and reflected some of the changes for sealing engine as well.